### PR TITLE
Adds sortagrad for ASR/LM/TTS

### PR DESCRIPTION
**Pull request by gtache**
Sends the samples from smallest to largest during the first epoch.

### CODE REVIEW

**Comment by gtache** [comment]
I'll check that everything is working, but now I have the dump_pcm tests failing for some reason (same for the gru_support branch)

**Comment by sw005320** [comment]
> I'll check that everything is working, but now I have the dump_pcm tests failing for some reason (same for the gru_support branch)

@kamo-naoyuki, could you take a look at it?

@gtache, did you observe some benefits of sortgrad? (convergence speed and/or performance)

**Comment by kamo-naoyuki** [comment]
@gtache Sorry, this is a bug of kaldiio and now fixed.

**Comment by gtache** [comment]
@kamo-naoyuki Thanks!
@sw005320 Well the project I'm working on will end in approximately one month and all my (4) gpus are allocated for this, so I can't test now (except on something like an4), and I haven't tried comparing to be honest.    
I imagine the difference should be more noticeable in a large dataset with high disparities in the sample lengths. If someone has the spare power to test I imagine something like Tedlium would be a good candidate.

**Comment by sw005320** [comment]
OK. Then, for now, could you make it `false`? After I review it, I'll merge it and test it with several benchmarks.

**Comment by gtache** [comment]
@sw005320 Oh yes of course, I'm just testing to see if everything is working fine.

**Comment by gtache** [comment]
Ok it should be good now

**Comment by sw005320** [comment]
@hirofumi0810, could you review this PR?
Also, I remember that you also have the experience. Please give us your comments.

**Comment by gtache** [comment]
Strange, I remember that in the papers I've read they only do it in the first epoch. Well that just means more customization and comparison possibilities 😄

**Comment by sw005320** [comment]
> Strange, I remember that in the papers I've read they only do it in the first epoch. Well that just means more customization and comparison possibilities 😄

I had several experience on the sortagrad. This seemed to provide slightly better convergence, but the final performance was comparable to the normal one, and I could not convince it but this is long time ago. I'll redo it with several other benchmarks once we merge it. @hirofumi0810 has different experience as he mentioned above.

**Comment by sw005320** [comment]
can you add a brief (one sentence) explanation of the sortagrad option in `run.sh` (the meaning of 0, -1, and other number)?

**Comment by gtache** [comment]
@sw005320 It seems that the second travis configuration (conda=false, python=2.7, pytorch=1.0.0) is consistently slower than the others (3-5x) and that can cause builds to timeout.    
Perhaps that should be tested with a recipe for example to see if this slowdown is reproducible. I've talked with a friend/colleague and he got a big slowdown when updating to Pytorch 1.0, and it seems that it happens to several other people (just search for "Pytorch 1.0 slow").

**Comment by sw005320** [comment]
> @sw005320 It seems that the second travis configuration (conda=false, python=2.7, pytorch=1.0.0) is consistently slower than the others (3-5x) and that can cause builds to timeout.
> Perhaps that should be tested with a recipe for example to see if this slowdown is reproducible. I've talked with a friend/colleague and he got a big slowdown when updating to Pytorch 1.0, and it seems that it happens to several other people (just search for "Pytorch 1.0 slow").

Thanks, I'll check it. @kamo-naoyuki, @kan-bayashi, please let me know if you also observe this. It is no problem for us to keep to use pytorch 0.4.X for a while.

**Comment by kamo-naoyuki** [comment]
I see. It is slow surely. Now travis is too heavily testing, so how about disabling the branch of  `USE_CONDA=false`?

File: espnet/tts/pytorch_backend/tts.py [hunk 1]
```diff
@@ -274,30 +278,34 @@ def train(args):
...
                                    args.minibatches, args.batch_sort_key,
-                                   min_batch_size=args.ngpu if args.ngpu > 1 else 1)
+                                   min_batch_size=args.ngpu if args.ngpu > 1 else 1, shortest_first=use_sortagrad)
     # hack to make batchsize argument as 1
     # actual bathsize is included in a list
     if args.n_iter_processes > 0:
-        train_iter = chainer.iterators.MultiprocessIterator(
+        train_iter = ToggleableShufflingMultiprocessIterator(
             TransformDataset(train_batchset, converter.transform),
-            batch_size=1, n_processes=args.n_iter_processes, n_prefetch=8, maxtasksperchild=20)
```
Comment (kamo-naoyuki) [inline espnet/tts/pytorch_backend/tts.py]: @gtache ToggleableShufflingMultiprocessIterator doesn't have `maxtasksperchild`, please add it.

**Pull request merged**